File: CPP/ChromaFilter.cpp
#include "ChromaFilter.h"

ChromaFilter::ChromaFilter(AVObject s) 
{
	setAVSource (s);
}

ChromaFilter::ChromaFilter(AVObject s, int convertMode) 
{
	this->setAVSource (s);
	this->setChromaSamplingFromY4M(convertMode);
}

void ChromaFilter::setAVSource (AVObject s)
{
	source = s;
	
	this->setFrameRate(s.getFrameRate());
	this->setSampleAspect(s.getSampleAspect());
	this->setHeight(s.getHeight());
	this->setWidth(s.getWidth());
	this->setInterlaced(s.getIsInterlaced());
	this->setInterlaceTopFieldFirst(s.getInterlaceTopFieldFirst());
}	

void ChromaFilter::setAVChroma(PixelFormat convertMode)
{

	this->setChromaSampling(convertMode);
	this->allocFrame();

	imgConvertCtx = sws_getContext(this->getWidth(), this->getHeight(), source.getChromaSampling(), 
								   this->getWidth(), this->getHeight(), convertMode, SWS_BICUBIC, NULL, NULL, NULL); 

}

void ChromaFilter::setY4MChroma(int convertMode)
{
	
	this->setChromaSamplingFromY4M(convertMode);
	this->allocFrame();
	
	imgConvertCtx = sws_getContext(this->getWidth(), this->getHeight(), source.getChromaSampling(), 
								   this->getWidth(), this->getHeight(), this->getChromaSampling(), SWS_BICUBIC, NULL, NULL, NULL); 
	
}

int ChromaFilter::decodeNextFrame(void)
{

	AVFrame *pFrameSource;
	
	source.decodeNextFrame();
	pFrameSource = source.getAVFrame();
	
	return sws_scale(imgConvertCtx,  (const uint8_t * const *)((AVPicture *)pFrameSource)->data,
					 pFrameSource->linesize, 0, this->getHeight(),pFrame->data, pFrame->linesize);
	
}

void ChromaFilter::dumpFormat(void)
{
	std::cerr << "chroma conversion filter: " << this->getChromaSamplingName() << "\n";
	source.dumpFormat();
}

File: objective_c/chromaFilter.m

#import "chromaFilter.h"

@implementation chromaFilter

- (id)initWithAVObject:(AVObject *)s
{
	
	[self init];

	source = s;
	[self setFrameRate:[s getFrameRate]];
	[self setSampleAspect:[s getSampleAspect]];
	[self setHeight:[s getHeight]];
	[self setWidth:[s getWidth]];
	[self setInterlaced:[s getIsInterlaced]];
	[self setInterlaceTopFieldFirst:[s getInterlaceTopFieldFirst]];
	
	return self;

}

- (id)initWithAVObject:(AVObject *)s toAVChroma:(int)convertMode
{

	if ([self initWithAVObject:s] != nil) 
	{
	
		[self setChromaSampling:convertMode];
	 
		[self allocFrame];
	 
		imgConvertCtx = sws_getContext([self getWidth], [self getHeight], [s getChromaSampling], 
									   [self getWidth], [self getHeight], convertMode, SWS_BICUBIC, NULL, NULL, NULL); 


		return self;
	}
	return nil;
}

- (id)initWithAVObject:(AVObject *)s toY4MChroma:(int)convertMode
{

	// need a chroma conversion function
	//	[self initWithAVObject:s toAVChroma: setChromaSamplingFromY4M:convertMode];
	
	// there should be a better way than this code duplication
	[self initWithAVObject:s];
	
	[self setChromaSamplingFromY4M:convertMode];
	
	[self allocFrame];
	
	imgConvertCtx = sws_getContext([self getWidth], [self getHeight], [s getChromaSampling], 
								   [self getWidth], [self getHeight], [self getChromaSampling], SWS_BICUBIC, NULL, NULL, NULL); 
	
	
	return self;
	
	
}

-(int)decodeNextFrame
{

	AVFrame *pFrameSource;
	
	[source decodeNextFrame];
	
	pFrameSource = [source getAVFrame];
	return sws_scale(imgConvertCtx,  (const uint8_t * const *)((AVPicture *)pFrameSource)->data,
					 pFrameSource->linesize, 0, [self getHeight],pFrame->data, pFrame->linesize);
	
}

- (void)dumpFormat
{
	NSLog(@"chroma conversion filter: %@\n",[self getChromaSamplingName]);
	[source dumpFormat];
}


@end